cosima_cookbook/database.py

```diff
@@ -578,86 +578,93 @@ def build_index(
-            prune_files(expt, session, missing_files_objs, delete=(prune == "delete"))
+        # find all netCDF files in the experiment directory
+        files = find_files(directory, followsymlinks=followsymlinks)
 
         if force:
-            # Delete entries that we're reindexing
-            for f in files & existing_files:
-                existing_ncfile = (
-                    session.query(NCFile)
-                    .join(NCExperiment)
-                    .filter(NCExperiment.experiment == str(directory.name))
-                    .filter(NCExperiment.root_dir == str(directory.resolve()))
-                    .filter(NCFile.ncfile == f)
-                    .one_or_none()
-                )
-                if existing_ncfile is not None:
-                    print("Re-indexing {}".format(f))
-                    session.delete(existing_ncfile)
-            session.flush()
+            # Prune all files to force re-indexing data
+            _prune_files(expt, session, {}, delete=True)
         else:
-            # Only pass files that are not already in DB
-            files = files - existing_files
+            # Prune files that exist in the database but are not present on disk
+            _prune_files(expt, session, files, delete=(prune == "delete"))
 
-        indexed += index_experiment(files, session, expt, client)
+        if len(files) > 0:
+            if len(expt.ncfiles) > 0:
+                # Only pass files that are not already in DB
+                files = {
+                    f.ncfile
+                    for f in session.query(NCFile)
+                    .with_parent(expt)
+                    .filter(NCFile.ncfile.notin_(files))
+                }
 
-        # if everything went smoothly, commit these changes to the database
-        session.commit()
+            indexed += index_experiment(files, session, expt, client)
+
+            # if everything went smoothly, commit these changes to the database
+            session.commit()
 
     return indexed
 
 
-def prune_files(expt, session, files, delete=True):
-    """Delete or mark as not present the database entries for specified
-    files objects within the given experiment
+def _prune_files(expt, session, files, delete=True):
+    """Delete or mark as not present the database entries that are
+    not present in the list of files
     """
-    for f in files:
+    if len(expt.ncfiles) == 0:
+        # No entries in DB, special case just return as there is nothing
+        # to prune and trying to do so will raise errors
+        return
+
+    # Missing are physically missing from disk, or where marked as not
+    # present previously. Can also be a broken file which didn't index
+    missing_ncfiles = (
+        session.query(NCFile)
+        .with_parent(expt)
+        .filter(NCFile.ncfile.notin_(files) | (NCFile.present == False))
+    )
+    if missing_ncfiles.first() is not None:
```

Comment:
Is this needed? Since `missing_ncfiles` is an un-executed query, it's fine if it affects 0 rows. (Also applies to the `len` check above)

Reply:
I put in those tests because I was getting errors like this:
```python
E                   sqlalchemy.exc.StatementError: (sqlalchemy.exc.InvalidRequestError) Can't resolve value for column experiments.id on object <NCExperiment at 0x14980124cdc0>; no value has been set for this column
E                   [SQL: SELECT ncfiles.ncfile AS ncfiles_ncfile 
E                   FROM ncfiles 
E                   WHERE ? = ncfiles.experiment_id AND ncfiles.ncfile NOT IN (?)]
E                   [parameters: [immutabledict({})]]

/home/502/aph502/.local/lib/python3.9/site-packages/sqlalchemy/orm/relationships.py:1772: StatementError
```
I just commented out that check and immediately got the same error in this test:

https://github.com/COSIMA/cosima-cookbook/blob/c23b089571ec19624ab91f6b1c2544c0f61e7d0b/test/test_indexing.py#L86

At this location.

Reply:
You're correct, this check isn't required for tests to pass, so I'll remove it. I suppose it was an abundance of caution from the other ones that do trip up you've highlighted elsewhere. Tah.

Reply:
Actually I think I also didn't want to do a `commit` if there hadn't been any changes, but whatevs.